### PR TITLE
Fix for Puppi MET and MET significance compatibility between AOD and miniAOD (80X), rebased

### PR DESCRIPTION
**Pull request by mmarionncern**
Previous version was #16174, rebased on top of 80X.
Planned for 80X miniAODv2

Packages touched :
- PhysicsTools/PatAlgos
- PhysicsTools/PatUtils
- RecoMET/METAlgorithms
- RecoMET/METProducers

Changes :
- fix scheduled mode (do not affect unscheduled mode)
- change the constructor of MET in the RecoMET extractor to copy properly the sumET from the slimmedMET collection when updating a miniAOD
- MET significance computation at the AOD level updated to match the miniAOD level
- Update of the test cfg files
- fix the puppi JEC and the use of slimmedJetPuppi when recomputing the puppiMET from miniAOD
- modify the MET signifiance code to match by reference the pfCandidates to the jets/lepton constituents + use of an extra recovery by delta"4Vector" matching when reference matching is failing

Expected changes in data file contents :
- small changes in the mET signifiance when producing miniAOD
- puppi MET values when rerunning on miniAOD
- storage of the sumEt when recomputing MET on top of miniAOD

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoMET/METFilters/test/test_badChargedCandidateFilter.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+process.load('Configuration.StandardSequences.Services_cff')
+process.load('RecoMET.METFilters.BadChargedCandidateFilter_cfi')
+process.load('RecoMET.METFilters.BadChargedCandidateSummer16Filter_cfi')
+
+## for miniAOD running
+process.BadChargedCandidateFilter.muons = cms.InputTag("slimmedMuons")
+process.BadChargedCandidateFilter.PFCandidates = cms.InputTag("packedPFCandidates")
+
+process.BadChargedCandidateFilter.debug = cms.bool(True)
```
Comment (slava77) [inline RecoMET/METFilters/test/test_badChargedCandidateFilter.py]: "debug" was removed from BadParticleFilter.
Better drop these lines from the test configs as well

File: RecoMET/METFilters/test/test_badMuonCandidateFilter.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+## for miniAOD running
+process.BadPFMuonFilter.muons = cms.InputTag("slimmedMuons")
+process.BadPFMuonFilter.PFCandidates = cms.InputTag("packedPFCandidates")
+
+## for miniAOD running ICHEP
+process.BadPFMuonSummer16Filter.muons = cms.InputTag("slimmedMuons")
+process.BadPFMuonSummer16Filter.PFCandidates = cms.InputTag("packedPFCandidates")
+
+process.BadPFMuonFilter.debug = cms.bool(True)
+process.BadPFMuonSummer16Filter.debug = cms.bool(True)
```
Comment (slava77) [inline RecoMET/METFilters/test/test_badMuonCandidateFilter.py]: as in the other test file: these lines have no effect and were removed in the same configs in 9X.
Please drop here as well to have a consistent backport

File: PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py [hunk 3]
```diff
@@ -1387,7 +1425,7 @@ def ak4JetReclustering(self,process, pfCandCollection, patMetModuleSequence, pos
                         bottomCollection = cms.InputTag("tmpPFCandCollPtr"+postfix) )
                         )
                 pfCandColl = cms.InputTag("pfNoPileUpJME"+postfix)
-
```
Comment (slava77) [inline PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py]: The part right after "pfCHS = cms.EDFilter" was implemented differently in the earlier version of this PR
(the pfCHS object was manipulated and than was used to override pfNoPileUpJME+postfix after the ``` if self._parameters["onMiniAOD"] ``` conditional)

Looking at 9X implementation, which is most similar in context lines , it seems like the following is missing for the onMiniAOD case
```
                setattr(process,"pfNoPileUpJME"+postfix,pfCHS)
                pfCandColl = cms.InputTag("pfNoPileUpJME"+postfix)
```

File: PhysicsTools/PatAlgos/python/slimming/puppiForMET_cff.py [hunk 4]
```diff
@@ -39,8 +40,19 @@ def makePuppiesFromMiniAOD( process ):
...
-    #Line below doesn't work because of an issue with references in MiniAOD without setting useRefs=>False and using delta R
-    #process.puppiForMET.puppiCandName    = 'puppiMerged'
-    #process.puppiForMET.useRefs          = False
+    #Line below replaces reference linking wiht delta R matching because the puppi references after merging are not consistent with those of the original packed candidate collection
+    process.puppiForMET.useRefs          = False
+    #Line below points puppi MET to puppi no lepton which increases the response
+    process.puppiForMET.puppiCandName    = 'puppiMerged'
+    #Avoid recomputing the weights available in MiniAOD
+    #process.puppi.useExistingWeights = True
+    #process.puppiNoLep.useExistingWeights = True
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/puppiForMET_cff.py]: 9X version has this set to true.
Why are these commented out for this backport?

File: CommonTools/PileupAlgos/plugins/PuppiPhoton.cc [hunk 5]
```diff
@@ -31,17 +31,19 @@ PuppiPhoton::PuppiPhoton(const edm::ParameterSet& iConfig) {
   tokenPFCandidates_     = consumes<CandidateView>(iConfig.getParameter<edm::InputTag>("candName"));
   tokenPuppiCandidates_  = consumes<CandidateView>(iConfig.getParameter<edm::InputTag>("puppiCandName"));
   tokenPhotonCandidates_ = consumes<CandidateView>(iConfig.getParameter<edm::InputTag>("photonName"));
-  tokenPhotonId_         = consumes<edm::ValueMap<bool>  >(iConfig.getParameter<edm::InputTag>("photonId")); 
+  reco2pf_               = mayConsume<edm::ValueMap<std::vector<reco::PFCandidateRef> > >(iConfig.getParameter<edm::InputTag>("recoToPFMap"));
+  tokenPhotonId_         = mayConsume<edm::ValueMap<bool>  >(iConfig.getParameter<edm::InputTag>("photonId"));
```
Comment (slava77) [inline CommonTools/PileupAlgos/plugins/PuppiPhoton.cc]: the implementation is different from what's in 9X.
I think that this part can be implemented in the same way (it also avoids mayConsume by using consumes)

File: CommonTools/PileupAlgos/plugins/PuppiPhoton.cc [hunk 6]
```diff
@@ -74,25 +79,32 @@ void PuppiPhoton::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
...
-      for( const edm::Ref<pat::PackedCandidateCollection> & ref : pPho->associatedPackedPFCandidates() ) {
-	if(fabs(pfCol->ptrAt(ref.key())->eta()) < eta_ ) {
-	  phoIndx.push_back(ref.key());
-	  phoCands.push_back(&(*(pfCol->ptrAt(ref.key()))));
-	}
+    if(runOnMiniAOD_) {
+      const pat::Photon *pPho = dynamic_cast<const pat::Photon*>(&(*itPho));
+      if(pPho != 0) {
+        for( const edm::Ref<pat::PackedCandidateCollection> & ref : pPho->associatedPackedPFCandidates() ) {
+	  if(fabs(pfCol->ptrAt(ref.key())->eta()) < eta_ ) {
```
Comment (slava77) [inline CommonTools/PileupAlgos/plugins/PuppiPhoton.cc]: please used
```  	  if(fabs(ref->eta()) < eta_ ) {  ``` as in 9X

File: CommonTools/PileupAlgos/plugins/PuppiPhoton.cc [hunk 7]
```diff
@@ -74,25 +79,32 @@ void PuppiPhoton::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
...
+	    phoIndx.push_back(ref.key());
+	    phoCands.push_back(&(*(pfCol->ptrAt(ref.key()))));
+	  }
+        }
+        continue;
+      }
+      const pat::Electron *pElectron = dynamic_cast<const pat::Electron*>(&(*itPho));
+      if(pElectron != 0) {
+        for( const edm::Ref<pat::PackedCandidateCollection> & ref : pElectron->associatedPackedPFCandidates() )
+	  if(fabs(pfCol->ptrAt(ref.key())->eta()) < eta_ )  {
```
Comment (slava77) [inline CommonTools/PileupAlgos/plugins/PuppiPhoton.cc]: same ```if(fabs(ref->eta()) < eta_ ) {``` here and also below

File: CommonTools/PileupAlgos/plugins/PuppiPhoton.cc [hunk 8]
```diff
@@ -105,7 +117,7 @@ void PuppiPhoton::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
   std::vector<float> lWeights;
```
Comment (slava77) [inline CommonTools/PileupAlgos/plugins/PuppiPhoton.cc]: 9X also had at L112 (see #17072) :
```diff
-  iEvent.getByToken(tokenWeights_,pupWeights);
+  if(useValueMap_)
+    iEvent.getByToken(tokenWeights_,pupWeights);
```

**Comment by slava77** [comment]
since miniAOD v2 may actually be done in 9X, until that is decided, it is not very important to update this PR.

**Comment by slava77** [comment]
ping

**Comment by mmarionncern** [comment]
@slava77 I plan to work on it on Thursday/Friday, I cannot do it sooner unfortunately

File: PhysicsTools/PatAlgos/python/slimming/puppiForMET_cff.py (hunk 4 above)
Comment (mmarionncern) [inline PhysicsTools/PatAlgos/python/slimming/puppiForMET_cff.py]: This is related to the puppi tuning that is not the same between the RECO step and the miniAOD step. As the backport now contains the proper tuning, I turned on those lines

File: CommonTools/PileupAlgos/plugins/PuppiPhoton.cc (hunk 8 above)
Comment (mmarionncern) [inline CommonTools/PileupAlgos/plugins/PuppiPhoton.cc]: I updated the 80X version with the 91X code, the backport is in sync now

File: PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py (hunk 3 above)
Comment (mmarionncern) [inline PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py]: Corrected

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/18341	92f73cd
- updates to miniAOD MET variables: modified regular MET significance, changed computation of PUPPI MET. All are in the scope of legacy re-miniAOD v2 (in case it happens in 80X). Changes files are consistent with what we have in 9X/master.
- jenkins tests pass and comparisons with baseline show differences in PUPPI MET (all variables) and in slimmedMET significance. All are expected.

I will put this PR on hold to indicate that it is supposed to be merged only when we are ready to build a re-miniAOD v2 release.

**Comment by slava77** [comment]
hold

to indicate that it is supposed to be merged only when we are ready to build a re-miniAOD v2 release.

**Comment by ahinzmann** [comment]
@slava77 We'd like to backport also #19587 for 80X. Should we merge it into this PR, or what is the appropriate way to get this into 80X?

**Comment by slava77** [comment]
On 8/2/17 3:02 AM, ahinzmann wrote:
> @slava77 <https://github.com/slava77> We'd like to backport also #19587
> <https://github.com/cms-sw/cmssw/pull/19587> for 80X. Should we merge it
> into this PR, or what is the appropriate way to get this into 80X?

I think that it is practical to merge so that you have just one PR for
recipes.
Given that it changes default miniAOD outputs, it is possible that this
PR will never get merged in 80X.

**Comment by zdemirag** [comment]
@slava77 : I am not sure I follow your comment on that it is possible this PR will never be merged in 80X. The statement was "to indicate that it is supposed to be merged only when we are ready to build a re-miniAOD v2 release."

These changes are needed for the legacy re-reco, re-miniaod v2 release.

**Comment by slava77** [comment]
On 8/3/17 8:52 AM, Zeynep Demiragli wrote:
> @slava77 <https://github.com/slava77> : I am not sure I follow your
> comment on that it is possible this PR will never be merged in 80X. The
> statement was "to indicate that it is supposed to be merged only when we
> are ready to build a re-miniAOD v2 release."
> 
> These changes are needed for the legacy re-reco, re-miniaod v2 release.

currently there is no plan to re-miniAOD v2 release in 80X.
The plan is to do it in 9X (perhaps 93X or 94X, whichever version is
there in time for that).

**Comment by slava77** [comment]
@mmarionncern @zdemirag 
The plan is fairly clear now to not make re-miniAOD of the legacy 80X/2016 with 80X.

Please close this PR or clarify the reasons why it should stay open.

Thank you.

**Comment by gouskos** [comment]
Hi @slava77, This means that fixes in this are not included in the 9X versions? Thanks.

**Comment by mmarionncern** [comment]
If 80X is not used, the PR can be closed

**Comment by slava77** [comment]
On 2/1/18 6:49 AM, gouskos wrote:
> Hi @slava77 <https://github.com/slava77>, This means that fixes in this
> are not included in the 9X versions? Thanks.
> 

I think that everything useful is already included in 9X since a while.
This PR is/was a backport for 8X.

**Pull request closed without merging**